routes.go

```diff
@@ -1,15 +1,41 @@
 package main
 
 import (
-	//"fmt"
+	"fmt"
+	"github.com/goamz/goamz/s3"
+	"io"
 	"log"
 	"net/http"
 	"net/url"
 	"path"
+	"strconv"
+	"time"
 )
 
+var httpClient = &http.Client{}
+
+const MAX_SOURCE_PULL_WAIT = time.Minute
+const MAX_WAIT_HEADER = "x-max-wait-duration"
+
 type Routes struct {
-	config *ProxyConfig
+	config   *ProxyConfig
+	requests requestMutex
+}
+
+func NewRoutes(config *ProxyConfig) Routes {
+	return Routes{
+		config:   config,
+		requests: newRequestMutex(),
+	}
+}
+
+func (self *Routes) constructKeyName(reqUrl *url.URL) string {
+	key := path.Join(self.config.Prefix, reqUrl.Path)
+	// Strip any starting slashes out of the bucket path...
+	if key[0:1] == "/" {
+		key = key[1:]
+	}
```

Comment:
We don't have any keys that starts with slash... But they are allowed in S3. Therefore it's best to check that the `prefix` does not start with a slash at start up, and either crash or show a warning in this case. Doing it here might modify the reqUrl.Path. If it's necessary to strip a slash from `reqUrl.Path` (I suspect it might be), we should always do this.

Again, I don't think it causes us problems here (because of our S3 bucket structure), but in general when working with S3. We can't just strip/fold slashes.